nomad/periodic.go

```diff
@@ -0,0 +1,529 @@
+		default:
+		}
+	}
+
+	p.logger.Printf("[DEBUG] nomad.periodic: deregistered periodic job %q", jobID)
+	return nil
+}
+
+// ForceRun causes the periodic job to be evaluated immediately.
+func (p *PeriodicDispatch) ForceRun(jobID string) error {
+	p.l.Lock()
+	defer p.l.Unlock()
+
+	// Do nothing if not enabled
+	if !p.enabled {
+		return fmt.Errorf("periodic dispatch disabled")
+	}
+
+	job, tracked := p.tracked[jobID]
+	if !tracked {
+		return fmt.Errorf("can't force run non-tracked job %v", jobID)
+	}
+
+	return p.createEval(job, time.Now())
+}
+
+// run is a long-lived function that waits til a job's periodic spec is met and
+// then creates an evaluation to run the job.
+func (p *PeriodicDispatch) run() {
+	defer close(p.waitCh)
+
+	// Do nothing if not enabled.
+	p.l.RLock()
+	if !p.enabled {
+		p.l.RUnlock()
+		return
+	}
+	p.l.RUnlock()
+
+	var now time.Time
+
+PICK:
+	// If there is nothing wait for an update.
+	p.l.RLock()
+	if p.heap.Length() == 0 {
+		p.l.RUnlock()
+		p.logger.Printf("[DEBUG] nomad.periodic: no periodic jobs; waiting")
+		select {
+		case <-p.stopCh:
+			return
+		case <-p.updateCh:
+		}
+		p.l.RLock()
+	}
+
+	nextJob, err := p.heap.Peek()
+	p.l.RUnlock()
+	if err != nil {
+		select {
+		case <-p.stopCh:
+			return
+		default:
+			p.logger.Printf("[ERR] nomad.periodic: failed to determine next periodic job: %v", err)
+			return
+		}
+	}
+
+	launchTime := nextJob.next
+
+	// If there are only invalid times, wait for an update.
+	if launchTime.IsZero() {
+		p.logger.Printf("[DEBUG] nomad.periodic: job %q has no valid launch time", nextJob.job.ID)
+		select {
+		case <-p.stopCh:
+			return
+		case <-p.updateCh:
+			goto PICK
+		}
+	}
+
+	now = time.Now()
+	p.logger.Printf("[DEBUG] nomad.periodic: launching job %q in %s",
+		nextJob.job.ID, nextJob.next.Sub(now))
+
+	select {
+	case <-p.stopCh:
+		return
+	case <-p.updateCh:
+		goto PICK
+	case <-time.After(nextJob.next.Sub(now)):
+		// Get the current time so that we don't miss any jobs will we are
+		// creating evals.
+		nowUpdate := time.Now()
+
+		// Create evals for all the jobs with the same launch time.
+		p.l.Lock()
+		for {
+			if p.heap.Length() == 0 {
+				break
+			}
+
+			j, err := p.heap.Peek()
+			if err != nil {
+				p.logger.Printf("[ERR] nomad.periodic: failed to determine next periodic job: %v", err)
+				break
+			}
+
+			if j.next != launchTime {
+				break
+			}
+
+			if err := p.heap.Update(j.job, j.job.Periodic.Next(nowUpdate)); err != nil {
+				p.logger.Printf("[ERR] nomad.periodic: failed to update next launch of periodic job %q: %v", j.job.ID, err)
+			}
+
+			p.logger.Printf("[DEBUG] nomad.periodic: launching job %v at %v", j.job.ID, launchTime)
+			go p.createEval(j.job, launchTime)
+		}
+
+		p.l.Unlock()
+		now = nowUpdate
+	}
+
+	goto PICK
+}
+
+// createEval instantiates a job based on the passed periodic job and submits an
+// evaluation for it.
+func (p *PeriodicDispatch) createEval(periodicJob *structs.Job, time time.Time) error {
```

Comment:
This is the logic you want to abstract out into an interface to more easily test both separately